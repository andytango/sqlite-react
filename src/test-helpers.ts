import { initDb } from "./init";
import { DbOpts } from "./types";
import Worker from "web-worker";

export const dbOpts: DbOpts = {
  sqlDataUrl: "src/test-db.sqlite",
  sqlJsWorkerPath: "src/test-worker.js",
  getDbFile,
};

async function getDbFile() {
  return new ArrayBuffer(0);
}

export function getTestWorker() {
  return new Worker(dbOpts.sqlJsWorkerPath);
}

export function createTestDbManager() {
  const worker = getTestWorker();
  const fn = jest.spyOn(worker, "postMessage");
  const dbInit = () => initDb({ ...dbOpts, worker });

  return { worker, dbInit, fn };
}
