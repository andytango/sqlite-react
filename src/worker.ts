import Worker from "web-worker";
import { DbOpts, DbResponse } from "./types";

export type DbWorkerFactory = typeof createDbWorker;
export type DbWorker = ReturnType<DbWorkerFactory>;

type DbWorkerState = {
  worker: Worker;
  isReady: boolean;
  initQueue: PendingActions;
  terminateQueue: PendingActions;
  isTerminating: boolean;
  didTerminate: boolean;
};

type PendingActions = (() => Promise<void>)[];

export function createDbWorker(opts: DbOpts) {
  const state = createInitialState(opts);

  async function init() {
    return initDbWithCallbacks(state, opts);
  }

  async function exec(sql: string) {
    if (state.didTerminate) {
      throw new Error("[DB Worker] Cannot exec, worker was terminated");
    }

    return execWorker(state, sql);
  }

  async function terminate() {
    if (state.isTerminating) {
      throw new Error("[DB Worker] Already terminating");
    }

    if (state.didTerminate) {
      throw new Error("[DB Worker] Already terminated");
    }

    state.isTerminating = true;
    for (let i = 0; i < state.terminateQueue.length; i++) {
      await state.terminateQueue[i]();
    }
    state.terminateQueue = [];
    terminateWorker(state);
    state.isTerminating = false;
    state.didTerminate = true;
  }

  return { init, exec, terminate };
}

function createInitialState(opts: DbOpts) {
  return {
    worker: initWebWorker(opts),
    isReady: false,
    initQueue: [],
    isTerminating: false,
    didTerminate: false,
    terminateQueue: [],
  } as DbWorkerState;
}

async function initDbWithCallbacks(state: DbWorkerState, opts: DbOpts) {
  await openDbFile(state, opts);

  for (let i = 0; i < state.initQueue.length; i++) {
    await state.initQueue[i]();
  }

  state.isReady = true;
  state.initQueue = [];
}

function initWebWorker(opts: DbOpts) {
  const worker = opts.webWorker || new Worker(opts.sqlJsWorkerPath);
  addEventListeners(worker);

  return worker;
}

function addEventListeners(worker: Worker) {
  worker.addEventListener("error", handleError);
}

function handleError(e: ErrorEvent) {
  console.error("[DB-WORKER] Error:", e.error);
}

async function openDbFile(state: DbWorkerState, opts: DbOpts) {
  const { sqlDataUrl, getDbFile = fetchDbFile } = opts;
  const buffer: ArrayBuffer = await getDbFile(sqlDataUrl);
  return performAction(state, "open", { buffer });
}

async function fetchDbFile(url: string) {
  const res = await fetch(url);
  return res.arrayBuffer();
}

type WorkerActionPayload = { sql: string } | { buffer: ArrayBuffer };

function execWorker(state: DbWorkerState, sql: string) {
  if (state.isReady) {
    return performQueryAction(state, sql);
  }

  return new Promise<DbResponse>((resolve) => {
    state.initQueue.push(async () => {
      const res = await performQueryAction(state, sql);
      resolve(res);
    });
  });
}

function performQueryAction(state: DbWorkerState, sql: string) {
  return performAction(state, "exec", { sql });
}

function performAction(
  state: DbWorkerState,
  action: string,
  payload: WorkerActionPayload
) {
  const { worker, terminateQueue } = state;
  const id = getNewMessageId();
  const message = { id, action, ...payload };

  return new Promise((resolve, reject) => {
    const handleTerminate = async () => {
      resolve({ type: "abort", id });
    };

    terminateQueue.push(handleTerminate);

    const handleResponse = (event: MessageEvent) => {
      if (event.data.id === id) {
        worker.removeEventListener("message", handleResponse);
        removePendingAction(terminateQueue, handleTerminate);
        resolve({ type: "result", ...event.data });
      }
    };

    worker.addEventListener("message", handleResponse);
    worker.postMessage(message);

    setTimeout(() => {
      worker.removeEventListener("message", handleResponse);
      console.error(`[DB-WORKER] Message timeout`, message);
      reject({ type: "error", error: "[DB-WORKER] Message timeout", id });
    }, 30e3);
  }) as Promise<DbResponse>;
}

let messageId = 0;

function removePendingAction(
  pending: PendingActions,
  action: () => Promise<void>
) {
  pending.splice(pending.indexOf(action));
}

function getNewMessageId() {
  return messageId++;
}

function terminateWorker(state: DbWorkerState) {
  const { worker } = state;
  worker.removeEventListener("error", handleError);
  worker.terminate();
  state.didTerminate = true;
}
