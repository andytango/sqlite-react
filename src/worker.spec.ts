import Worker from "web-worker";
import { createTestDbWorker, dbOpts } from "./test-helpers";
import { createDbWorker } from "./worker";

describe("createDbWorker", () => {
  it("adds event listeners to the worker", async () => {
    const webWorker = new Worker(dbOpts.sqlJsWorkerPath);
    const fn = jest.spyOn(webWorker, "addEventListener");

    createDbWorker({ ...dbOpts, webWorker });

    expect(fn).toHaveBeenCalledWith("error", expect.any(Function));

    webWorker.terminate();
  });

  it("terminates worker and aborts queries", async () => {
    const webWorker = new Worker(dbOpts.sqlJsWorkerPath);
    const fn = jest.spyOn(webWorker, "terminate");

    const dbWorker = createDbWorker({ ...dbOpts, webWorker });
    await dbWorker.init();

    const promise = dbWorker.exec("select example_col from example_slow_table");
    await dbWorker.terminate();

    expect(fn).toHaveBeenCalled();

    const res = await promise;
    expect(res).toEqual({ type: "abort", id: expect.any(Number) });

    await expect(dbWorker.exec("some val")).rejects.toEqual(
      new Error("[DB Worker] Cannot exec, worker was terminated")
    );
  });

  it("throws error if terminate called while already terminating", async () => {
    const webWorker = new Worker(dbOpts.sqlJsWorkerPath);

    const dbWorker = createDbWorker({ ...dbOpts, webWorker });
    await dbWorker.init();

    const promise1 = dbWorker.exec(
      "select example_col from example_slow_table"
    );
    const promise2 = dbWorker.terminate();

    expect(dbWorker.terminate()).rejects.toEqual(
      new Error("[DB Worker] Already terminating")
    );

    await promise1;
    await promise2;
  });

  it("throws error if terminate called after worker has terminated", async () => {
    const webWorker = new Worker(dbOpts.sqlJsWorkerPath);

    const dbWorker = createDbWorker({ ...dbOpts, webWorker });
    await dbWorker.init();

    const promise1 = dbWorker.exec(
      "select example_col from example_slow_table"
    );
    const promise2 = dbWorker.terminate();

    await promise1;
    await promise2;

    expect(dbWorker.terminate()).rejects.toEqual(
      new Error("[DB Worker] Already terminated")
    );
  });

  it("executes sql via worker", async () => {
    const { webWorker, dbWorker } = createTestDbWorker();
    const fn1 = jest.spyOn(webWorker, "postMessage");
    const fn2 = jest.spyOn(webWorker, "addEventListener");

    await dbWorker.init();
    const res = await dbWorker.exec("select 1 as val");

    expect(fn1).toHaveBeenCalledWith({
      id: expect.any(Number),
      action: "exec",
      sql: "select 1 as val",
    });

    expect(fn2).toHaveBeenCalledWith("message", expect.any(Function));

    expect(res).toEqual({
      type: "result",
      id: expect.any(Number),
      results: [{ columns: ["val"], values: [[1]] }],
    });

    dbWorker.terminate();
  });

  it("opens sqlite file via worker", async () => {
    const { webWorker, dbWorker } = createTestDbWorker();
    const fn1 = jest.spyOn(webWorker, "postMessage");
    await dbWorker.init();

    expect(fn1).toHaveBeenCalledWith({
      id: expect.any(Number),
      action: "open",
      buffer: expect.any(ArrayBuffer),
    });

    dbWorker.terminate();
  });

  it("queues up queries will initialising", async () => {
    const { webWorker, dbWorker } = createTestDbWorker();
    const fn1 = jest.spyOn(webWorker, "postMessage");

    await Promise.all([dbWorker.exec("select 1 as val"), dbWorker.init()]);

    expect(fn1).toHaveBeenNthCalledWith(1, {
      id: expect.any(Number),
      action: "open",
      buffer: expect.any(ArrayBuffer),
    });

    expect(fn1).toHaveBeenNthCalledWith(2, {
      id: expect.any(Number),
      action: "exec",
      sql: "select 1 as val",
    });

    dbWorker.terminate();
  });
});
