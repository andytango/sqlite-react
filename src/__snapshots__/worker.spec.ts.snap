// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createDbWorker executes sql via worker 1`] = `
Object {
  "id": 0,
  "results": Array [
    Object {
      "columns": Array [
        "val",
      ],
      "values": Array [
        Array [
          1,
        ],
      ],
    },
  ],
}
`;
